File: iosApp/Models/GameState.swift
import SwiftUI

public enum GameStatus: String {
    case ready
    case playing
    case paused
    case gameOver
}

public struct GameState {
    public var status: GameStatus = .ready
    public var score: Int = 0
    public var highScore: Int = 0
    public var survivalSeconds: CGFloat = 0
    public var asteroidsDodged: Int = 0
    public var powerUpsCollected: Int = 0
    public var shieldActive: Bool = false
    public var shieldRemainingSeconds: CGFloat = 0
    public var shieldMaxSeconds: CGFloat = 8.0
    public var player: PlayerShip = PlayerShip()
    public var asteroids: [Asteroid] = []
    public var powerUps: [PowerUp] = []
    public var particles: [Particle] = []
    public var floatingTexts: [FloatingText] = []
    public var stars: [Star] = []
    public var speedMultiplier: CGFloat = 1.0
    public var waveLevel: Int = 1

    public init(stars: [Star] = []) {
        self.stars = stars
    }

    public var formattedSurvivalTime: String {
        let total = Int(survivalSeconds)
        let minutes = total / 60
        let seconds = total % 60
        return String(format: "%02d:%02d", minutes, seconds)
    }

    public var shieldRatio: CGFloat {
        if shieldMaxSeconds > 0 {
            return min(max(shieldRemainingSeconds / shieldMaxSeconds, 0.0), 1.0)
        }
        return 0.0
    }
}

File: app/src/main/java/com/example/spacedodger/model/GameState.kt
package com.example.spacedodger.model

/**
 * High-level lifecycle states of the game session.
 */
enum class GameStatus {
    READY,
    PLAYING,
    PAUSED,
    GAME_OVER
}

/**
 * Immutable snapshot of the Space Dodger game state.
 * Designed for predictable state flow in Compose Multiplatform.
 */
data class GameState(
    val status: GameStatus = GameStatus.READY,
    val score: Int = 0,
    val highScore: Int = 0,
    val survivalSeconds: Float = 0f,
    val asteroidsDodged: Int = 0,
    val powerUpsCollected: Int = 0,
    val shieldActive: Boolean = false,
    val shieldRemainingSeconds: Float = 0f,
    val shieldMaxSeconds: Float = 8f,
    val player: PlayerShip = PlayerShip(),
    val asteroids: List<Asteroid> = emptyList(),
    val powerUps: List<PowerUp> = emptyList(),
    val particles: List<Particle> = emptyList(),
    val floatingTexts: List<FloatingText> = emptyList(),
    val stars: List<Star> = emptyList(),
    val speedMultiplier: Float = 1.0f,
    val waveLevel: Int = 1
) {
    /**
     * Formatted survival time "MM:SS"
     */
    val formattedSurvivalTime: String
        get() {
            val totalSec = survivalSeconds.toInt()
            val minutes = totalSec / 60
            val seconds = totalSec % 60
            return "%02d:%02d".format(minutes, seconds)
        }

    /**
     * Progress ratio (0.0f..1.0f) of the active shield power-up.
     */
    val shieldRatio: Float
        get() = if (shieldMaxSeconds > 0f) (shieldRemainingSeconds / shieldMaxSeconds).coerceIn(0f, 1f) else 0f
}
